samples/server.c

```diff
@@ -208,21 +209,46 @@ static void do_dma_io(vfu_ctx_t *vfu_ctx, struct server_data *server_data)
 
     assert(vfu_ctx != NULL);
 
+    struct iovec iov = {0};
+
     /* Write some data, chunked into multiple calls to exercise offsets. */
     for (int i = 0; i < count; ++i) {
-        addr = server_data->regions[0].iova.iov_base + i * size;
+        addr = server_data->regions[region].iova.iov_base + i * size;
         ret = vfu_addr_to_sgl(vfu_ctx, (vfu_dma_addr_t)addr, size, sg, 1,
                               PROT_WRITE);
+                              
         if (ret < 0) {
             err(EXIT_FAILURE, "failed to map %p-%p", addr, addr + size - 1);
         }
 
         memset(&buf[i * size], 'A' + i, size);
-        vfu_log(vfu_ctx, LOG_DEBUG, "%s: WRITE addr %p size %d", __func__, addr,
-                size);
-        ret = vfu_sgl_write(vfu_ctx, sg, 1, &buf[i * size]);
-        if (ret < 0) {
-            err(EXIT_FAILURE, "vfu_sgl_write failed");
+
+        if (use_messages) {
+            vfu_log(vfu_ctx, LOG_DEBUG, "%s: MESSAGE WRITE addr %p size %d",
+                    __func__, addr, size);
+            ret = vfu_sgl_write(vfu_ctx, sg, 1, &buf[i * size]);
+            if (ret < 0) {
+                err(EXIT_FAILURE, "vfu_sgl_write failed");
+            }
+        } else {
+            vfu_log(vfu_ctx, LOG_DEBUG, "%s: DIRECT WRITE  addr %p size %d",
+                    __func__, addr, size);
+            ret = vfu_sgl_get(vfu_ctx, sg, &iov, 1, 0);
+            if (ret < 0) {
+                err(EXIT_FAILURE, "vfu_sgl_get failed");
+            }
+            assert(iov.iov_len == (size_t)size);
+            memcpy(iov.iov_base, &buf[i * size], size);
+
+            /*
+             * When directly writing to client memory we are responsible for
+             * tracking dirty pages. We assert that all dirty writes are within
+             * the first page of region 1. In fact, all regions are only one
+             * page in size.
+             */
+            vfu_sgl_mark_dirty(vfu_ctx, sg, 1);
```

Comment:
I think we shuold probably do vfu_sgl_put() here instead. Servers are supposed to match them up.

Reply:
I agree, but how do we then demonstrate `vfu_sgl_mark_dirty`? Or do we not need to?

Reply:
I think along with `vfu_sgl_put`, we should still have `vfu_sgl_mark_dirty` with a comment explaining that _in this example_ it isn't strictly necessary as it's already marked dirty by the former, and that it can be useful in cases where the server needs to mark guest memory dirty without releasing the mapping.

Reply:
Precisely